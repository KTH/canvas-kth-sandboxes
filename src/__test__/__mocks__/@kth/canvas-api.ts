type MockRes = {
    body : any
}

const _mockValues : {
    get: Record<string, MockRes>;
    request: Record<string, MockRes>;
    _result: Record<string, any>;
} =
{
    get: {},
    request: {},
    _result: {},
};

export class CanvasApiMock{
    static get(url:string, value:MockRes){
        if(_mockValues.get[url])
            throw new Error(`Mock value already exists for ${url}`);
        _mockValues.get[url] = value;
    }

    static request(url:string, _method:string, value:MockRes){
        if(_mockValues.request[url])
            throw new Error(`Mock value already exists for ${url}`);
        _mockValues.request[url] = value;
    }

    static result(){
        return _mockValues["_result"];
    }

        
}

export default class CanvasApi{

    _baseUrl:string = "";
    _token:string = "";
    constructor(url:string, token:string){
        this._baseUrl = url;
        this._token = token;

    }

    async get(url:string) {
        return _mockValues.get[url];
    }

    async request(url:string, method:string, params:Record<string, any> ){
        if (method !== "POST"){
            throw new Error(`This mock does not support ${method}`);
        }
        _mockValues["_result"][url] = params;
        return _mockValues.request[url];
    }
}