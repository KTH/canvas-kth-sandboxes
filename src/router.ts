
import {createCourse, enrollUser, getUser, getRole} from "./canvasApi"
import {Request, Response } from "express";
import { Router} from "express";
import log from "skog";
import path from "path";



const TEST_ACCOUNT_IDS = ["97021", "97017", "97016", "97018", "97020", "97019"];
const KTH_DEV_ID = 18;

const router = Router();

router.use("/public", homepage);
router.get("/public", (req, res) => {
  res.sendFile(path.join(__dirname, '/html/index.html'));
});
router.get("/_monitor", monitor);


async function homepage(req: Request, res: Response, next: Function){
  if(!await checkAuth(req, res)){
    res.redirect("/canvas-kth-sandboxes/auth");
  }else if(!await checkPermission(req, res)){
    res.status(403).json({message: "Permission denied, du saknar behörighet för den här appen."});
  }else{
    next();
  }
}

async function checkAuth(req: Request, res: Response):Promise<boolean>{
  if(!req.session.accessToken){
    return false;
  }
  return true;
}

async function checkPermission(req: Request, res:Response){

  if (!req.session.accessToken){
    return false;}

  const role = await getRole(req.session.accessToken);
  
  if (!role){
    return false;
  }
  if (!role.find(r => r.role_id = KTH_DEV_ID)){
     return false;
  }
  return true;
}


async function monitor(req: Request, res: Response) {  
  try {
    res.send("APPLICATION_STATUS: OK");
  } catch (error) {
    log.error("Error: something went wrong.");
    res.send("Application status: ERROR");
    
  }
};

router.post("/create-sandbox", createSandbox);


async function createSandbox(req: Request, res: Response): Promise<void> {
  const sisUserId = req.body.userId;
  const schoolId = req.body.school;

  const user = await getUser(sisUserId);
  if (!user){
    res.send("SIS Id does not exist");
    return;
  }
  const userName = user.body.login_id.split("@")[0];
  const userId = user.body.id;
  const course = await createCourse(userName, schoolId);
  const courseId = course.body.id;

  await enrollUser(userId, courseId, "TeacherEnrollment");

  for (const testStudent of TEST_ACCOUNT_IDS){
    await enrollUser(testStudent, courseId, "StudentEnrollment");
  }

  const htmlRes= `
  <!DOCTYPE html>  
  <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Sandbox for Canvas@kth</title>
    </head>
    <body>
        <h1 id="message">Sandbox have been created for ${userName}</h1>
        <p><a href="${process.env.CANVAS_API_URL}courses/${courseId}">URL to Sandbox</a></p>
    </body>
  </html>
  `

  res.send(htmlRes);

}

export default router;

// export for testing
export {
  checkAuth,
  createSandbox
}
