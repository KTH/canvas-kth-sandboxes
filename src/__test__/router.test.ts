import { describe, test } from "@jest/globals";
import { CanvasApiMock } from "./__mocks__/@kth/canvas-api";
import { createSandbox} from "../router";

CanvasApiMock.get("users/sis_user_id:1234/profile", { body: { login_id: "get_user_id@kth", id: "get_user_id" } });
CanvasApiMock.request("accounts/school_id/courses", "POST", { body: { id: "course_id" } });
for (let i = 0; i < 7; i++) {
  CanvasApiMock.request(`courses/user_id_${i}/enrollments`, "POST", { body: undefined });
}


describe("Testing logic for creating sandboxes", () => {


  test("testsuite works", () => { });


  test("CreateSandbox, input res and req output html", async () => {
    const accessToken = "access_token";
    const userId = "1234";
    const schoolId = "school_id";
    const response = await createSandbox(userId, schoolId, accessToken);
    expect(CanvasApiMock.result()).toMatchSnapshot();
    expect(response).toMatchSnapshot();


  });
});
