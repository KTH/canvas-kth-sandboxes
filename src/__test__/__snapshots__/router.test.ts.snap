// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing logic for creating sandboxes CreateSandbox, input res and req output html 1`] = `
{
  "accounts/school_id/courses": {
    "course": {
      "course_code": "Sandbox get_user_id",
      "name": "Sandbox get_user_id",
    },
  },
  "courses/course_id/enrollments": {
    "enrollment": {
      "enrollment_state": "active",
      "type": "StudentEnrollment",
      "user_id": "97019",
    },
  },
}
`;

exports[`Testing logic for creating sandboxes CreateSandbox, input res and req output html 2`] = `
"
  <!DOCTYPE html>  
  <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Sandbox for Canvas@kth</title>
    </head>
    <body>
        <h1 id="message">Sandbox have been created for get_user_id</h1>
        <p><a href="undefinedcourses/course_id">URL to Sandbox</a></p>
    </body>
  </html>
  <style>
  body{
    font-family: Arial, Helvetica, sans-serif;
    padding: 25px;
    font-color: black;
  }
  </style>
"
`;
